### PR TITLE
Add JSON serializers/deserializers for optimizers and gradients

### PR DESCRIPTION
**Pull request by bmhowe23**
## Description

`cudaq::optimizer` and `cudaq::gradient` are abstract virtual classes, so they need special handling for JSON serialization and deserialization. This PR adds that support.

These changes should be backwards compatible with the existing NVQC deployment as long as the new `std::optional<RestRequestOptFields> opt` field in the `RestRequest` is not populated. (And it is indeed not populated in any of our current code.)

This is all in support of remote VQE/optimizer execution on NVQC.

### CODE REVIEW

File: runtime/common/JsonConvert.h [hunk 1]
```diff
@@ -178,10 +197,236 @@ inline void from_json(const json &j, ExecutionContext &context) {
...
+  if (optimizer_type == OptimizerEnum::GRAD_DESC) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::gradient_descent>();
+    from_json(j, dynamic_cast<cudaq::optimizers::BaseEnsmallen &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (optimizer_type == OptimizerEnum::SGD) {
+    auto ret_ptr = std::make_unique<cudaq::optimizers::sgd>();
+    from_json(j, dynamic_cast<cudaq::optimizers::BaseEnsmallen &>(*ret_ptr));
+    return ret_ptr;
+  }
```
Comment (1tnguyen) [inline runtime/common/JsonConvert.h:327]: nit: is there a reason we don't use `switch case` here?

File: runtime/common/JsonConvert.h [hunk 2]
```diff
@@ -178,10 +197,236 @@ inline void from_json(const json &j, ExecutionContext &context) {
...
+    auto ret_ptr = std::make_unique<cudaq::gradients::forward_difference>();
+    from_json(j,
+              dynamic_cast<cudaq::gradients::forward_difference &>(*ret_ptr));
+    return ret_ptr;
+  }
+  if (gradient_type == GradientEnum::PARAMETER_SHIFT) {
+    auto ret_ptr = std::make_unique<cudaq::gradients::parameter_shift>();
+    from_json(j, dynamic_cast<cudaq::gradients::parameter_shift &>(*ret_ptr));
+    return ret_ptr;
+  }
```
Comment (1tnguyen) [inline runtime/common/JsonConvert.h:390]: same question as https://github.com/NVIDIA/cuda-quantum/pull/1800/files#r1635394341

**Review by 1tnguyen — APPROVED** [review]
LGTM 👍

File: runtime/common/JsonConvert.h (hunk 1 above)
Comment (bmhowe23) [reply runtime/common/JsonConvert.h]: Updated in e8a16e4db. I also removed the unnecessary dynamic_cast just to tidy it up.

File: runtime/common/JsonConvert.h (hunk 2 above)
Comment (bmhowe23) [reply runtime/common/JsonConvert.h]: Updated in e8a16e4db. I also removed the unnecessary dynamic_cast just to tidy it up.

**Pull request merged**